cloudmarker/alerts/emailalert.py

```diff
@@ -1,119 +1,40 @@
-        written in the email body
+        """Save event record in a buffer.
 
         Arguments:
-            record (dict): Data to send via email.
+            record (dict): An event record.
         """
         for _, value in record.items():
             self._stringbuffer.append(repr(value))
 
     def done(self):
-        """Perform final cleanup tasks.
-
-        This method is called after all records have been written. In
-        this example implementation, we properly terminate the JSON
-        array in the email body.
-
-        The connection is made based whether the SMTP or SMTP_SSL is
-        required which is determined by ``use_ssl`` param. The SMTP
-        login process is conducted if ``username`` param is provided.
-
-        """
-        message = MIMEMultipart()
-        message['Date'] = email.utils.formatdate(localtime=True)
-        message['Subject'] = self._subject
-        message['From'] = self._sender
-        message['To'] = email.utils.COMMASPACE.join(self._to)
-
-        # Incase if the buffer is empty then the default message will be sent
-        if self._stringbuffer:
-            self._body = ''.join(self._stringbuffer)
-        message.attach(MIMEText(self._body))
-
-        smtp = self._prepare_smtp_session()
-        try:
-            smtp.sendmail(self._sender, self._to, message.as_string())
-        except smtplib.SMTPException as e:
-            _log.error('Failed to send email: %s', e)
-        finally:
-            smtp.quit()
-
-    def _prepare_smtp_session(self):
-        """Return SMTP connection object.
-
-        Create a SMTP connection based on whether the SSL param.
-        If SSL connection is required (for eg: in case of gmail) then
-        method will return a SMTP_SSL connection. In other cases if the
-        authentication is not required or SMTP_SSL connection is not
-        required then a plain SMTP connection object is returned.
-        For gmail follow the steps given below even to work for 2 factor
-        authentication
-        1. Log-in into Gmail with your account
-        2. Navigate to https://security.google.com/settings/security/
-        apppasswords
-        3. In 'select app' choose 'custom', give it an arbitrary name and
-        press generate
-        4. It will give you 16 chars token.
-        5. Use that token as password for login
-        6. Host is smtp.gmail.com and port is 465
-
-        """
-        if self._use_ssl:
-            smtp = smtplib.SMTP_SSL(host=self._host, port=self._port)
-            smtp.set_debuglevel(True)
-        else:
-            smtp = smtplib.SMTP(host=self._host, port=self._port)
-            smtp.set_debuglevel(True)
-            smtp.starttls()
-
-        if self._username:
-            smtp.login(self._username, self._password)
-        return smtp
+        """Send the buffered events as an email alert."""
+        self._kwargs['content'] = '\n\n'.join(self._stringbuffer)
+        util.send_email(**self._kwargs)
```

Comment:
I believe this file is easier to review if we just take a look at the final code instead of the diffs. You can do that with `git checkout auditemails; vim cloudmarker/alerts/emailalert.py` or you can just see the following code which I have included for your convenience:

```python
"""Email alert plugin."""


import logging

from cloudmarker import util

# Define module-level logger.
_log = logging.getLogger(__name__)


class EmailAlert:
    """A plugin to send email alerts."""

    def __init__(self, **kwargs):
        """Create an instance of :class:`EmailAlert` plugin.

        This class accepts the same arguments as
        :func:`cloudmarker.util.send_email`.

        The ``content`` argument is not honoured. Even if a ``content``
        argument is provided, it is ignored by this class because this
        class defines its own content from the event records it receives
        in its :meth:`write` method.
        """
        self._kwargs = kwargs
        self._buffer = []

    def write(self, record):
        """Save event record in a buffer.

        Arguments:
            record (dict): An event record.
        """
        for _, value in record.items():
            self._buffer.append(repr(value))

    def done(self):
        """Send the buffered events as an email alert."""
        self._kwargs['content'] = '\n\n'.join(self._buffer)
        util.send_email(**self._kwargs)
```